### ISSUE TITLE
once-only events fired before custom.js is loaded

### ISSUE THREAD

**Issue by mpacer** [jupyter/notebook#2499]
Earlier in https://github.com/jupyter/notebook/issues/1730 we had an issue where the point at which `custom.js` was loaded was delayed so that standard triggers wouldn't work. It seems that with notebook 5.0 this bug was reintroduced.

This is a problem for scripts using `notebook_loaded.Notebook` as well as `app_initialized.NotebookApp` (likely the cause of https://github.com/jupyter/notebook/issues/2403). 

I don't know if https://github.com/jupyter/notebook/pull/1732 give an idea for a similar solution — but if it should be a Promise, my guess is it needs a slightly different kind of fix. Happy to attempt a fix if someone wants to point me in the right direction; seems like a good way to learn about Promises.

**Comment by mpacer** [comment]
this is the test `custom.js` that I was using to test stuff. 

```javascript
// file ~/custom/custom.js

console.log("hello1");

require([
	"base/js/namespace",
	"base/js/events"
], 
	function(Jupyter, events){
		console.log("hello2");	
		events.on('notebook_loaded.Notebook', function(){
			console.log("hello3");
		});
		events.on('app_initialized.NotebookApp', function(){
			console.log("hello4");
		});
});

console.log("hello5");
```

**Comment by rolweber** [comment]
You're right. From a brief look, I'd say that of all the changes in #1732, only these are still in place:

* `notebook/templates/page.html`
* `tools/build-main.js`

while these are missing:

* `notebook/static/edit/js/main.js`
* `notebook/static/notebook/js/main.js`
* `notebook/static/terminal/js/main.js`
* `notebook/static/tree/js/main.js`

You could work from the [file diff](https://github.com/jupyter/notebook/pull/1732/files) to re-apply the missing changes. If you have time to spend and/or want to learn git, you could also examine the history to see where they got lost :-)

**Comment by takluyver** [comment]
Looks like they were changed back in #1378, because we don't want a bad custom.js to prevent the notebook from loading.

I think the longer term fix is that rather than fiddling with the loading order, we should have promises for one-time events like `app_initialized.NotebookApp`, so that custom code and extensions can wait for something that may already have occurred.

**Comment by rolweber** [comment]
@mpacer: I suggest to rename this issue. Something like "once-only events fired before custom.js is loaded", maybe.

I'm not firm with Promises either, so the following are really just pointers. I like this introduction: https://developers.google.com/web/fundamentals/getting-started/primers/promises

For the `notebook_loaded.Notebook` event, you'll have to change [notebook/static/notebook/js/notebook.js](https://github.com/jupyter/notebook/blob/master/notebook/static/notebook/js/notebook.js).  
I would create a Promise in the constructor around line 100, so it's immediately available to `custom.js` code. Then resolve the promise when the `notebook_loaded.Notebook` event is fired, near line 3099, and reject it when the `notebook_load_failed.Notebook` event is fired, near line 3112.

The tricky thing is that to create a promise, you need to pass a function that expects two arguments: one function to resolve the promise, the other to reject it. In good style, you'd write the actual code for what's going to happen inside this function of yours. But due to lack of experience, I can't wrap my head around that yet. The hacky way out: use your function to store the two arguments in the Notebook object. Then, in the two places mentioned above, call one function or the other.  
Maybe someone else can give you (and me) better advice.

```js
// near line 100
this.notebook_ready = new Promise(function (resolve, reject) {
   // this is so insanely ugly...
   // (hope that 'this' is the same here as two lines above)
   this._resolve_notebook_ready = resolve;
   this._reject_notebook_ready = reject;
});

// near line 3099
// need to pass something, the notebook object is as good as any
this._resolve_notebook_ready(this);

// near line 3112
this._reject_notebook_ready(error);
```
In your `custom.js`, you'd use that promise like this:
```js

function(Jupyter, events){
  Jupyter.notebook.notebook_ready.then(function(){
    // notebook_loaded.Notebook must have been fired, or is about to be
    // Jupyter.notebook is passed as an argument and can be ignored
    console.log("hello3");
  });
}
```
I hope that gets you started.

**Comment by takluyver** [comment]
@Carreau should be able to give you advice on promises.

**Comment by ghazpar** [comment]
@takluyver But if the event app_initialized.NotebookApp arrives later, how to we do stuff that need to be donne before the notebook is loaded like, for instance, telling codemirror to display line numbers by default?

**Comment by Carreau** [comment]
I haven't dived into this code in a long time. 
Promises would help, though because of asynchronicity you may still be triggered too late. 
So if you are triggered after the notebook loads, you would still need for CodeMirror to set the default value, and loop over all the cells to apply the default to existing cells.

**Comment by ghazpar** [comment]
I don't understand why breaking the `app_initialized.NotebookApp` event is a better alternative than letting a "bad custon.js prevent the loading of a notebook"? If I think I need a custom.js for my app, than it is my responsibility to make sure that this js doesn't fail. Moreover, why should this bad custom.js issue require a change of timing? Can't it be handle through exceptions?

For the line numbers, after the fact, I can certainly tell each existing code cell to display them, but my app also needs to hide many of the notebook standard interface (menubar, some element of the toolbar, etc.) that I don't want my users to see and use. Currently (up to 4.4), I could do that before the interface appeared, using the `app_initialized.NotebookApp` event. In Notebook 5.0, the custom.js appears to be called after the notebook has finished loading, thus the standard interface first appears before being hidden. It may not by a show breaker, but is is not cool!

So I repeat my question: why is it useful at this point to break the timing behavior of the custom.js mechanism, especially if the plan is to eventually remove it in favor of extensions? Better to remove it now and force transition than to leave it and break it.

**Comment by takluyver** [comment]
It's a general principle that user configuration shouldn't prevent your entire application from working at all. It doesn't 'break' the timing behaviour, it changes it. Undoubtedly if we change it back, another group of people will complain about things that break.

We have to load it at a different time to catch the exception, I believe, because of the way Javascript module loading works. We make one `require()` call which loads multiple modules. If any of them fail, the whole load fails. So in order to catch exceptions from custom.js and carry on with the rest of the code, we need to separate loading that file from loading the other code.

**Comment by ghazpar** [comment]
I wonder which group of people you are referring to? Those that have developed a `custom.js` since the introduction of 5.0 and that assume it is called **after** the notebook has loaded and fired both 
`app_initialized.NotebookApp` and `notebook_loaded.Notebook` events? Or those that have a **broken** custom script and would thus complain that it prevents their notebook from running correctly? 

I may agree with the invoked general principle, but certainly question your use of the word "Undoubtedly". I further disagree with your assertion that changing timing is not breaking. It **does** break the `custom.js` example given in the documentation, as the two events are no longer received!

**Comment by ghazpar** [comment]
I have no mastery of the javascript module loading intricacies, but wouldn't it be possible to simply load a stub module that would itself load the custom module within a try/catch? Then there would be no need to change timing...

**Comment by takluyver** [comment]
That could work. Do you want to make a PR? The PRs linked above should point you to the code in question.

**Comment by ghazpar** [comment]
I wish I could, but I am a js novice with no PR experience on github. So not sure where to start...

**Comment by mpacer** [comment]
Happy to take a stab at fixing this next week at the biannual meeting — as suggested I could use a bit of guidance about how both to think about and to implement this.

**Comment by takluyver** [comment]
Discussing the implementation will be easier with someone who you can talk to in person, but this is how I think about it:

*Events*:
- Consumer listens for events using a method called something like `on` or `register`.
- The same event may fire multiple times, and consumers are called each time.
- In the JS event system we use, events are identified by a name, not an object; this is convenient, because you don't need an event to be defined before you subscribe to it.

*Promises* (/futures):
- Consumer waits for completion using a method called something like `then()`.
- The promise is only completed *once*, but the consumer code can tell if it's already fired, so code that needs to run after it is complete can be run reliably.
- Promises are objects, so you need the promise to be defined before your code can use it.

**Comment by hotlogical** [comment]
Is there a workaround for this ? I've just moved to python 3(.6) - with the latest anaconda3.

With the test case above I don't see the log messages "hello3" and "hello4"

My use case is that my custom.js is loading codemirror emacs key bindings and I can't live without them !!!!

Thanks.

**Comment by takluyver** [comment]
I've added a promise in #2710

**Comment by takluyver** [comment]
And we've just merged that, so from 5.1 you can use code like this in custom.js:

```javascript
define([
        'base/js/namespace',
        'base/js/promises'
     ], function(Jupyter, promises) {
         promises.app_initialized.then(function(appname) {
             if (appname === 'NotebookApp') {
                 do_stuff()
             }
         });
     });
```

So far, this fires for `NotebookApp` and `DashboardApp`, the same ones that already have app_initialized events.

Do we want to do anything else on this issue?

**Comment by ghazpar** [comment]
What about the old notebook_loaded.Notebook event?

**Comment by takluyver** [comment]
Yup, I guess that makes sense too. See #2719.

**Comment by takluyver** [comment]
Closing this as solved by adding promises in #2710 and #2719. I know it's not a great solution for the code already using the event, but I think it's a better API for the future.

**Comment by rlamorea** [comment]
So, it seems that because of this change, the instructions here:
http://testnb.readthedocs.io/en/latest/public_server.html#embedding-the-notebook-in-another-website

are no longer valid.  Custom.js is called _after_ the notebook links are created (which are also created using the `IPython._target` variable, not `Jupyter._target`) and as a result they are created as `target="_blank"` and thus break the single-tab functionality.

Seems likely the docs are out of date, but I don't see an alternative.
What's the workaround on this?
